src/surroundings_menu.cpp

```diff
@@ -0,0 +1,1160 @@
+        data.selected_entry = *data.filtered_list.begin();
+    } else {
+        std::advance( it, -page_scroll );
+        data.selected_entry = *it;
+    }
+}
+
+template<typename T>
+void surroundings_menu::select_page_down_generic( T &data )
+{
+    auto it = std::find( data.filtered_list.begin(), data.filtered_list.end(), data.selected_entry );
+    if( it == data.filtered_list.end() || it == data.filtered_list.end() - 1 ) {
+        data.selected_entry = *data.filtered_list.begin();
+    } else if( std::distance( it, data.filtered_list.end() ) <= page_scroll ) {
+        data.selected_entry = *data.filtered_list.rbegin();
+    } else {
+        std::advance( it, page_scroll );
+        data.selected_entry = *it;
+    }
+}
+
+void surroundings_menu::select_page_up()
+{
+    auto_scroll = true;
+    switch( selected_tab ) {
+        case surroundings_menu_tab_enum::items:
+            select_page_up_generic( item_data );
+            break;
+        case surroundings_menu_tab_enum::monsters:
+            select_page_up_generic( monster_data );
+            break;
+        case surroundings_menu_tab_enum::terfurn:
+            select_page_up_generic( terfurn_data );
+            break;
+        default:
+            debugmsg( "invalid tab selected" );
+    }
+}
+
+void surroundings_menu::select_page_down()
+{
+    auto_scroll = true;
+    switch( selected_tab ) {
+        case surroundings_menu_tab_enum::items:
+            select_page_down_generic( item_data );
+            break;
+        case surroundings_menu_tab_enum::monsters:
+            select_page_down_generic( monster_data );
+            break;
+        case surroundings_menu_tab_enum::terfurn:
+            select_page_down_generic( terfurn_data );
+            break;
+        default:
+            debugmsg( "invalid tab selected" );
+    }
+}
+
+void surroundings_menu::select_prev_internal()
+{
+    switch( selected_tab ) {
+        case surroundings_menu_tab_enum::items:
+            item_data.selected_entry->select_prev();
+            break;
+        case surroundings_menu_tab_enum::monsters:
+            monster_data.selected_entry->select_prev();
+            break;
+        case surroundings_menu_tab_enum::terfurn:
+            terfurn_data.selected_entry->select_prev();
+            break;
+        default:
+            debugmsg( "invalid tab selected" );
+    }
+}
+
+void surroundings_menu::select_next_internal()
+{
+    switch( selected_tab ) {
+        case surroundings_menu_tab_enum::items:
+            item_data.selected_entry->select_next();
+            break;
+        case surroundings_menu_tab_enum::monsters:
+            monster_data.selected_entry->select_next();
+            break;
+        case surroundings_menu_tab_enum::terfurn:
+            terfurn_data.selected_entry->select_next();
+            break;
+        default:
+            debugmsg( "invalid tab selected" );
+    }
```

Comment:
This and every other method here seem very verbose and duplicative. Perhaps something like this would be better:

```c++
   tab_data list;
   switch( selected_tab ) {
        case surroundings_menu_tab_enum::items:
            list = item_data;
            break;
        case surroundings_menu_tab_enum::monsters:
            list = monster_data;
            break;
        case surroundings_menu_tab_enum::terfurn:
            list = terfurn_data;
            break;
        default:
            debugmsg( "invalid tab selected" );
    }
    list.selected_entry->select_next();
```

They could be shortened further by putting the switch into a common method, perhaps called `get_currently_selected_list`. Then it would just be:

```c++
   get_currently_selected_list().selected_entry->select_next();
```

Reply:
Some of that can be simplified, yes. I did so with the filtering for example. But for some others it isn't that easy, because the tabs have different underlying data types (item, Creature, mapdata_common_t).